common/src/main/java/com/microsoft/identity/common/internal/controllers/BaseController.java

```diff
@@ -552,6 +562,48 @@ protected AccountRecord getCachedAccountRecord(
         return targetAccount;
     }
 
+    @Nullable
+    private AccountRecord getAccountWithFRTIfAvailable(@NonNull final SilentTokenCommandParameters parameters,
+                                                       @NonNull final MsalOAuth2TokenCache msalOAuth2TokenCache){
+
+        final String homeAccountId = parameters.getAccount().getHomeAccountId();
+        final String clientId = parameters.getClientId();
+
+        // check for FOCI tokens for the homeAccountId
+        final RefreshTokenRecord refreshTokenRecord = msalOAuth2TokenCache
+                .getFamilyRefreshTokenForHomeAccountId(homeAccountId);
+
+        if (refreshTokenRecord != null) {
+            try {
+                // foci token is available, make a request to service to see if the client id is FOCI and save the tokens
+                TokenCacheItemMigrationAdapter.tryFociTokenWithGivenClientId(
+                        parameters.getOAuth2TokenCache(),
+                        clientId,
+                        parameters.getRedirectUri(),
+                        refreshTokenRecord,
+                        parameters.getAccount()
+                );
```

Comment:
OK, we have the token then we make this call.  We stuff it into the datastore as a side effect, then go back to the datastore to pull it out again.  There's probably a reason we don't just return it... but anyway, this method returns a boolean value that we're not checking.  Should we check it?  Should we return the token instead?

Reply:
The use case here is:

We have a `Family Refresh Token` for an Microsoft first party client id(Teams for Life here) for this user after he signs in

Now we get a silent request for this `IAccount` with another Microsoft first party client id(Teams for Work) - at this point we don't have any account associated with this client id. 

What this change enables is, if the user has Family Refresh Token (applicable only for MSFT ), we try to make a `dummy` silent request in `tryFociTokenWithGivenClientId` with the FRT, default scopes and the new client id  and if the request succeeds we save the account and tokens for this client id in this cache. The access token here won't be used and will be replaced as it has just the default scopes.

Now the caller when it looks for the `IAccount` for this client, it gets one and makes the actual request for an access token for the required scopes passed in using the new RT which was saved. 

The boolean can be ignored in our case, as we have sufficient logging in `tryFociTokenWithGivenClientId` to understand if the request succeeds. 

Hope this helps

Reply:
Oh, no, thanks for bearing with me. 

OK, so the interesting side effect of tryFociWithGivenClientId is to bind the FRT to the new client id and store these refresh tokens by client id in the cache.  Is it possible to return a non-null AccountRecord from the call to getAccountByLocalAccountId even if the attempt to execute the family token refresh failed with the new client id? It seems like it shouldn't, right now, but that's largely because I'm thinking that the sole purpose of this method is to return an account iff this is a client id that part of the family, and if that was the case the refresh would have succeeded.

Reply:
yes, not just the side, i could say the primary purpose of this `tryFociWithGivenClientId` is to insert an `AccountRecord` for the given client, it's part of FOCI (we determine this by sending a request with FRT and if it succeeds).

And yes, if the token request with FRT failed , we should not get an `AccountRecord` when `getAccountByLocalAccountId` is invoked with that client id in this scenerio